### ISSUE TITLE
Adding GC and mappability with incompatible bin sizes massively duplicates CNV regions in the resulting `qseaSet`

### ISSUE THREAD

**Issue by lbeltrame** [cruk-mi/mesa#26]
The code attempts a `full join` here:

https://github.com/cruk-mi/mesa/blob/212a791cb4da2d2d2c099c4ef53b4b71d5f1ee65/R/combineQsets.R#L193

But, when joining large `qseaSets`, the result is too much for dplyr to handle:

```
r$> combineQsetsList(list(batch1, batch2, batch3, batch4))
No initial qseaSet given, using first element as initial qseaSet
Joining with `by = join_by(seqnames, start, end, width, strand, CpG_density)`                                                    
Joining with `by = join_by(seqnames, start, end, width, strand, CpG_density)`                                                    
Error in `dplyr::full_join()`:
! This join would result in more rows than dplyr can handle.
ℹ 50985959424 rows would be returned. 2147483647 rows is the maximum number allowed.                                             
ℹ Double check your join keys. This error commonly occurs due to a missing join key, or an improperly specified join condition.  
Run `rlang::last_trace()` to see where the error occurred.
```

**Comment by SPPearce** [comment]
Wow, what is going on there?
Something is not right, you shouldn't be getting anywhere near that many rows.
How big is your cnv slot of qseaSet2? What window size are you using?

**Comment by lbeltrame** [comment]
I'm currently at s conference but as soon as I get back (Monday) I'll let you know.
FTR these are normal samples.

**Comment by SPPearce** [comment]
`str(qseaSet)` is likely to be informative, if you can share that when you can.

**Comment by lbeltrame** [comment]
I noticed that there are **loads** of duplicated ranges in the CNV slot, is this normal?

```
unique(batch3@cnv)
GRanges object with 3021 ranges and 4 metadata columns:
         seqnames              ranges strand |   sampleE   sampleF   sampleG   sampleH
            <Rle>           <IRanges>  <Rle> | <numeric> <numeric> <numeric> <numeric>
     [1]     chr1           1-1000000      * |      0.09      0.05         0      0.05
     [2]     chr1     1000001-2000000      * |      0.09      0.05         0      0.05
     [3]     chr1     2000001-3000000      * |      0.09      0.05         0      0.05
     [4]     chr1     3000001-4000000      * |      0.09      0.05         0      0.05
     [5]     chr1     4000001-5000000      * |      0.09      0.05         0      0.05
     ...      ...                 ...    ... .       ...       ...       ...       ...
  [3017]     chrX 151000001-152000000      * |     -0.05     -0.15         0         0
  [3018]     chrX 152000001-153000000      * |     -0.05     -0.15         0         0
  [3019]     chrX 153000001-154000000      * |     -0.05     -0.15         0         0
  [3020]     chrX 154000001-155000000      * |     -0.05     -0.15         0         0
  [3021]     chrX 155000001-156000000      * |     -0.05     -0.15         0         0
  -------
  seqinfo: 23 sequences from an unspecified genome; no seqlengths
```

**Comment by SPPearce** [comment]
> I noticed that there are **loads** of duplicated ranges in the CNV slot, is this normal?

Nope, that is exactly what I suspected, it shouldn't have any duplicated regions there.
So we need to work out where they are coming from upstream.

**Comment by lbeltrame** [comment]
Where can I debug this? 
(In the mean time, I'll deduplicate the cnv slot to progress further)

**Comment by SPPearce** [comment]
> Where can I debug this? (In the mean time, I'll deduplicate the cnv slot to progress further)

Can you share with me your command that you used for `makeQset` please.

**Comment by SPPearce** [comment]
And what branch you are using I guess? If you are using a modified one with multiple of the fixes can you share that somewhere? @Steven-M-Hill is supposedly going to be looking at the PRs this week....

**Comment by lbeltrame** [comment]
Currently I run the `vignette` branch. I didn't yet switch to use a patched one (I wanted to review every change before charging in blindly...).

I used this command to make the qSet:

```
qseaSet <- makeQset(table BSgenome = "BSgenome.Hsapiens.UCSC.hg38", windowSize = 300, fragmentType = "cfDNA", CNVmethod = "HMMdefault", chrSelect = c(paste0("chr",1:22), "chrX"), hmmCopyGC = xt_gc_hg38_500kb, hmmCopyMap = xt_map_hg38)

```
The `xt_*` GC and mappability are simply the same objects as in mesa, but in UCSC notation. `Table` includes samples and their corresponding input.

**Comment by SPPearce** [comment]
And your `xt_gc_hg38_500kb` and `xt_map_hg38` objects, how many rows/columns are they?
xt_gc_hg38_500kb suggests it is a 500kb size, not the 1Mb default...

**Comment by lbeltrame** [comment]
There are about 6K rows for each object. Using the 500kbp variant may be have been an oversight as that's the bin size I use for CNV analysis on ctDNA. To make sure this wasn't the cause I'm running an analysis with the 1Mbp variants. I'll update once done.

I added `chr` fairly naively there with

`gc_hg38_1000kb[, `chr` := paste0("chr", `chr`)]`

**Comment by lbeltrame** [comment]
.... Well, it looks there aren't duplicates now. So I guess it was an incompatible bin size: Quite a silent error though because it went through unnoticed. I assume there's no way to detect it short of checking the size of the first range?

**Comment by SPPearce** [comment]
Yeah, need to add some checks for that then.
I'll edit it.
If you want to use 500kb bins, there is no reason you can't, just change the `CNVwindowSize` parameter.

**Comment by SPPearce** [comment]
Fixed by #29
